pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -651,6 +669,139 @@ func (s *DiscoveryServer) pushEds(push *model.PushContext, con *XdsConnection,
+	// A new ClusterLoadAssignment to return that will have both local and
+	// remote gateways (if any)
+	filteredCLA := &xdsapi.ClusterLoadAssignment{
+		ClusterName: cla.ClusterName,
+		Endpoints:   []endpoint.LocalityLbEndpoints{},
+		Policy:      cla.Policy,
+	}
+
+	// Weight (number of endpoints) for the EDS cluster for each remote networks
+	remoteEps := map[string]uint32{}
+
+	// Go through all cluster endpoints and add those with the same network as the sidecar
+	// to the result. Also count the number of endpoints per each remote network while
+	// iterating so that it can be used as the weight for the gateway endpoint
+	for _, ep := range cla.Endpoints {
+		onlyLocalLbEndpoints := []endpoint.LbEndpoint{}
+		foundRemote := false
+		for _, lbEp := range ep.LbEndpoints {
+			epNetwork := ""
+			if lbEp.Metadata != nil && lbEp.Metadata.FilterMetadata["istio"] != nil &&
+				lbEp.Metadata.FilterMetadata["istio"].Fields != nil &&
+				lbEp.Metadata.FilterMetadata["istio"].Fields["network"] != nil {
+				epNetwork = lbEp.Metadata.FilterMetadata["istio"].Fields["network"].GetStringValue()
+			}
+			if epNetwork == conNetwork {
+				// This is a local endpoint
+				onlyLocalLbEndpoints = append(onlyLocalLbEndpoints, lbEp)
+			} else {
+				// Remote endpoint. Increase the weight counter
+				remoteEps[epNetwork]++
+				foundRemote = true
+			}
+		}
+
+		if !foundRemote {
+			// This LocalityLbEndpoints has no remote endpoints so just
+			// add it to the result
+			filteredCLA.Endpoints = append(filteredCLA.Endpoints, ep)
+		} else {
+			// This LocalityLbEndpoints has remote endpoint so add to the result
+			// a new one that holds only local endpoints
+			newEp := endpoint.LocalityLbEndpoints{
+				Locality:            ep.Locality,
+				LbEndpoints:         onlyLocalLbEndpoints,
+				LoadBalancingWeight: ep.LoadBalancingWeight,
+				Priority:            ep.Priority,
+			}
+			filteredCLA.Endpoints = append(filteredCLA.Endpoints, newEp)
+		}
+	}
+
+	// If there is no MeshNetworks configuration, we don't have gateways information
+	// so just return the filtered results (with no remote endpoints)
+	if env.MeshNetworks == nil {
+		return filteredCLA
+	}
+
+	// Iterate over all networks that have the cluster endpoint (weight>0) and
+	// for each one of those add a new endpoint that points to the network's
+	// gateway with the relevant weight
+	for n, w := range remoteEps {
+		networkConf, found := env.MeshNetworks.Networks[n]
+		if !found {
+			continue
+		}
+		gws := networkConf.Gateways
+		if len(gws) == 0 {
+			continue
+		}
+
+		// There may be multiple gateways for the network. Add an LbEndpoint for
+		// each one of them
+		gwEndpoints := []endpoint.LbEndpoint{}
+		for _, gw := range gws {
+			// If the gateway address is DNS we can't add it to the endpoints
+			if gwIP := net.ParseIP(gw.GetAddress()); gwIP != nil {
+				addr := util.BuildAddress(gw.GetAddress(), gw.Port)
+				gwEp := endpoint.LbEndpoint{
+					Endpoint: &endpoint.Endpoint{
```

Comment:
Endpoint address allows both IPs and DNS records (per endpoint, actually) - does CDS control whether DNS is used for the entire set of endpoints, or can it vary per endpoint? [I couldn't tell from the Envoy docs](https://github.com/envoyproxy/envoy/blob/7f4c6fff2419623e88e799fd39da7871c6ac3399/api/envoy/api/v2/core/address.proto#L38-L42). If not then supporting DNS is blocked on updating CDS, RDS too I take it?

Reply:
In a conversation I had with @rshriram he pointed out that endpoints can't get DNS so at the moment I just filtered out those.

Reply:
👍